File: mailblastr-python/mailblastr/domains.py
"""Domains resource — create, verify, claim, and one-click DNS providers."""

from typing import Dict, TypedDict

from . import http_client
from .http_client import build_query, paginate, path_escape as _e

# Regions a domain can be created in today.
AVAILABLE_REGIONS = ("us-east-1", "ap-south-1")


class CreateParams(TypedDict, total=False):
    name: str  # required, at least two DNS labels
    region: str  # one of AVAILABLE_REGIONS; defaults to us-east-1
    custom_return_path: str  # MAIL FROM subdomain (Return-Path); defaults to 'send'
    open_tracking: bool  # default False
    click_tracking: bool  # default True
    custom_tracking: bool  # serve tracking links from your own subdomain
    tracking_subdomain: str  # custom tracking host label, e.g. 'email' (implies custom_tracking)
    tls: str  # 'opportunistic' | 'enforced'
    receiving: bool  # accept inbound mail (or use capabilities.receiving)
    capabilities: Dict[str, str]  # e.g. {"receiving": "enabled"}


class UpdateParams(TypedDict, total=False):
    open_tracking: bool
    click_tracking: bool
    tracking_subdomain: str
    custom_tracking: bool
    custom_return_path: str  # single DNS label, e.g. 'send'
    tls: str  # 'opportunistic' | 'enforced'
    receiving: bool
    capabilities: Dict[str, str]


class ClaimParams(TypedDict, total=False):
    name: str  # required
    region: str


class Domains:
    """``mailblastr.Domains`` — the /domains endpoints."""

    CreateParams = CreateParams
    UpdateParams = UpdateParams
    ClaimParams = ClaimParams
    AVAILABLE_REGIONS = AVAILABLE_REGIONS

    @classmethod
    def create(cls, params):
        """Add a sending domain. POST /domains"""
        return http_client.request("POST", "/domains", params)

    @classmethod
    def get(cls, domain_id):
        """Retrieve a domain (with its DNS records). GET /domains/:id"""
        return http_client.request("GET", f"/domains/{_e(domain_id)}")

    @classmethod
    def list(cls, params=None):
        """List domains. Called with no pagination params every domain is
        returned; pending claims (status ``claim``) are never listed here —
        read them with :meth:`get_claim`. GET /domains"""
        return http_client.request("GET", f"/domains{paginate(params)}")

    @classmethod
    def mx_check(cls, name):
        """Look up a domain's live MX records and whether they all point at
        MailBlastr — ``{"has_mx", "ours", "records"}``. Fails open: a DNS
        failure reports no records rather than raising.
        GET /domains/mx-check?name=..."""
        return http_client.request("GET", f"/domains/mx-check{build_query({'name': name})}")

    @classmethod
    def records_csv(cls, domain_id):
        """Download a domain's DNS records as CSV ``bytes`` (not JSON) — handy
        for handing to whoever runs your DNS.
        GET /domains/:id/records.csv"""
        return http_client.request_raw("GET", f"/domains/{_e(domain_id)}/records.csv")

    @classmethod
    def update(cls, domain_id, params):
        """Update domain settings. PATCH /domains/:id"""
        return http_client.request("PATCH", f"/domains/{_e(domain_id)}", params)

    @classmethod
    def verify(cls, domain_id):
        """Trigger DNS verification. POST /domains/:id/verify"""
        return http_client.request("POST", f"/domains/{_e(domain_id)}/verify")

    @classmethod
    def claim(cls, params):
        """Claim a domain already verified elsewhere. POST /domains/claim"""
        return http_client.request("POST", "/domains/claim", params)

    @classmethod
    def get_claim(cls, domain_id):
        """Retrieve a domain's claim record. GET /domains/:id/claim"""
        return http_client.request("GET", f"/domains/{_e(domain_id)}/claim")

    @classmethod
    def verify_claim(cls, domain_id):
        """Verify a domain claim. POST /domains/:id/claim/verify"""
        return http_client.request("POST", f"/domains/{_e(domain_id)}/claim/verify")

    @classmethod
    def detect_dns(cls, domain_id):
        """Detect the domain's DNS provider and available one-click apply
        methods. GET /domains/:id/dns/detect"""
        return http_client.request("GET", f"/domains/{_e(domain_id)}/dns/detect")

    @classmethod
    def apply_cloudflare_dns(cls, domain_id, params):
        """Apply this domain's DNS records via the Cloudflare API, then
        auto-verify. POST /domains/:id/dns/cloudflare — ``{"token": ...}``"""
        return http_client.request("POST", f"/domains/{_e(domain_id)}/dns/cloudflare", params)

    @classmethod
    def apply_godaddy_dns(cls, domain_id, params):
        """Apply this domain's DNS records via the GoDaddy API, then
        auto-verify. POST /domains/:id/dns/godaddy — ``{"key", "secret"}``"""
        return http_client.request("POST", f"/domains/{_e(domain_id)}/dns/godaddy", params)

    @classmethod
    def apply_namecheap_dns(cls, domain_id, params):
        """Apply this domain's DNS records via the Namecheap API (existing
        records preserved), then auto-verify.
        POST /domains/:id/dns/namecheap — ``{"apiUser", "apiKey", "userName"?}``"""
        return http_client.request("POST", f"/domains/{_e(domain_id)}/dns/namecheap", params)

    @classmethod
    def remove(cls, domain_id):
        """Delete a domain. DELETE /domains/:id"""
        return http_client.request("DELETE", f"/domains/{_e(domain_id)}")

File: mailblastr-java/src/main/java/com/mailblastr/resources/Domains.java
package com.mailblastr.resources;

import com.mailblastr.ListParams;
import com.mailblastr.MailblastrResponse;
import com.mailblastr.http.ApiClient;
import com.mailblastr.http.Query;
import com.mailblastr.requests.ClaimDomainRequest;
import com.mailblastr.requests.CreateDomainRequest;
import com.mailblastr.requests.UpdateDomainRequest;

import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Map;

/** Sending/receiving domains — {@code mailblastr.domains()}. */
public final class Domains extends Resource {
    public Domains(ApiClient api) { super(api); }

    /** {@code POST /domains} */
    public MailblastrResponse create(CreateDomainRequest request) {
        return api.request("POST", "/domains", request);
    }

    /** {@code GET /domains/:id} */
    public MailblastrResponse get(String id) {
        return api.request("GET", "/domains/" + enc(id));
    }

    /**
     * List domains. With no pagination params the route returns ALL of them
     * and {@code has_more:false}. Rows still awaiting a DNS-TXT ownership
     * claim ({@code status: "claim"}) are excluded — reach those through
     * {@link #getClaim(String)}. {@code GET /domains}
     */
    public MailblastrResponse list() { return list(null); }

    public MailblastrResponse list(ListParams params) {
        return api.request("GET", "/domains" + paginate(params));
    }

    /**
     * Check a domain's live MX records before enabling inbound.
     * Returns {@code { has_mx, ours, records }}; DNS failures fail open with
     * {@code has_mx:false}. {@code GET /domains/mx-check?name=}
     */
    public MailblastrResponse mxCheck(String name) {
        Query q = new Query().add("name", name);
        return api.request("GET", "/domains/mx-check" + q);
    }

    /**
     * Download the domain's DNS records as CSV text (the route returns
     * {@code text/csv}, not JSON). {@code GET /domains/:id/records.csv}
     */
    public String recordsCsv(String id) {
        byte[] bytes = api.requestRaw("GET", "/domains/" + enc(id) + "/records.csv");
        return new String(bytes, StandardCharsets.UTF_8);
    }

    /** Returns the slim ack {@code { object: 'domain', id }}. {@code PATCH /domains/:id} */
    public MailblastrResponse update(String id, UpdateDomainRequest request) {
        return api.request("PATCH", "/domains/" + enc(id), request);
    }

    /** Trigger DNS verification. {@code POST /domains/:id/verify} */
    public MailblastrResponse verify(String id) {
        return api.request("POST", "/domains/" + enc(id) + "/verify");
    }

    /** Claim a domain already verified elsewhere. {@code POST /domains/claim} */
    public MailblastrResponse claim(ClaimDomainRequest request) {
        return api.request("POST", "/domains/claim", request);
    }

    /** Retrieve a domain's claim record. {@code GET /domains/:id/claim} */
    public MailblastrResponse getClaim(String id) {
        return api.request("GET", "/domains/" + enc(id) + "/claim");
    }

    /** Verify a domain claim. {@code POST /domains/:id/claim/verify} */
    public MailblastrResponse verifyClaim(String id) {
        return api.request("POST", "/domains/" + enc(id) + "/claim/verify");
    }

    /**
     * Detect the domain's DNS provider and the one-click apply methods
     * available. {@code GET /domains/:id/dns/detect}
     */
    public MailblastrResponse detectDns(String id) {
        return api.request("GET", "/domains/" + enc(id) + "/dns/detect");
    }

    /** Apply DNS records via the Cloudflare API, then auto-verify. {@code POST /domains/:id/dns/cloudflare} */
    public MailblastrResponse applyCloudflareDns(String id, String token) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("token", token);
        return api.request("POST", "/domains/" + enc(id) + "/dns/cloudflare", body);
    }

    /** Apply DNS records via the GoDaddy API, then auto-verify. {@code POST /domains/:id/dns/godaddy} */
    public MailblastrResponse applyGoDaddyDns(String id, String key, String secret) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("key", key);
        body.put("secret", secret);
        return api.request("POST", "/domains/" + enc(id) + "/dns/godaddy", body);
    }

    /**
     * Apply DNS records via the Namecheap API (existing records preserved),
     * then auto-verify. Namecheap must have the calling server's IP
     * whitelisted. {@code POST /domains/:id/dns/namecheap}
     */
    public MailblastrResponse applyNamecheapDns(String id, String apiUser, String apiKey) {
        return applyNamecheapDns(id, apiUser, apiKey, null);
    }

    public MailblastrResponse applyNamecheapDns(String id, String apiUser, String apiKey, String userName) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("apiUser", apiUser);
        body.put("apiKey", apiKey);
        if (userName != null) body.put("userName", userName);
        return api.request("POST", "/domains/" + enc(id) + "/dns/namecheap", body);
    }

    /** {@code DELETE /domains/:id} */
    public MailblastrResponse remove(String id) {
        return api.request("DELETE", "/domains/" + enc(id));
    }
}
